internal/resources/azure/sql_managed_instance.go

```diff
@@ -0,0 +1,196 @@
+//
+// The azurerm_sql_managed_instance resource is deprecated in version 3.0 of the AzureRM provider and will be removed in version 4.0.
+// Please use the azurerm_mssql_managed_instance resource instead when available in infracost
+//
+// Only support for Gen5 is available on that resource
+//
+// More resource information here: https://azure.microsoft.com/en-gb/products/azure-sql/managed-instance/
+// Pricing information here: https://azure.microsoft.com/en-gb/pricing/details/azure-sql-managed-instance/single/
+type SQLManagedInstance struct {
+	Address            string
+	Region             string
+	SKU                string
+	LicenceType        string
+	Cores              *int64
+	StorageSizeInGb    *int64
+	StorageAccountType string
+	// LongTermRetentionStorageGB defines a usage param that allows users to define how many gb of cold storage the database uses.
+	// This is storage that can be kept for up to 10 years.
+	LongTermRetentionStorageGB *int64 `infracost_usage:"long_term_retention_storage_gb"`
+	BackupStorageGb            *int64 `infracost_usage:"backup_storage_gb"`
+}
+
+// PopulateUsage parses the u schema.UsageData into the SQLManagedInstance.
+// It uses the `infracost_usage` struct tags to populate data into the SQLManagedInstance.
+func (r *SQLManagedInstance) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+// BuildResource builds a schema.Resource from a valid SQLManagedInstance struct.
+// This method is called after the resource is initialised by an IaC provider.
+// See providers folder for more information.
+func (r *SQLManagedInstance) BuildResource() *schema.Resource {
+	costComponents := r.costComponents()
+
+	return &schema.Resource{
+		Name: r.Address,
+		UsageSchema: []*schema.UsageItem{
+			{Key: "backup_storage_gb", DefaultValue: 0, ValueType: schema.Int64},
+			{Key: "long_term_retention_storage_gb", DefaultValue: 0, ValueType: schema.Int64},
+		},
+		CostComponents: costComponents,
+	}
+}
+
+func (r *SQLManagedInstance) costComponents() []*schema.CostComponent {
+
+	costComponents := []*schema.CostComponent{
+		{
+			Name:           fmt.Sprintf("Compute (%s %s Cores)", strings.ToTitle(r.SKU), strconv.FormatInt(*r.Cores, 10)),
+			Unit:           "hours",
+			UnitMultiplier: decimal.NewFromInt(1),
+			HourlyQuantity: decimalPtr(decimal.NewFromInt(1)),
+			ProductFilter: &schema.ProductFilter{
+				VendorName:    strPtr(vendorName),
+				Region:        strPtr(r.Region),
+				Service:       strPtr(sqlMIServiceName),
+				ProductFamily: strPtr(sqlMIProductFamily),
+				AttributeFilters: ([]*schema.AttributeFilter{
+					{Key: "productName", Value: r.productDescription()},
+					{Key: "skuName", Value: r.meteredName()},
+				}),
+			},
+			PriceFilter: priceFilterConsumption,
+		},
+	}
+
+	if r.BackupStorageGb != nil {
+		costComponents = append(costComponents, r.sqlMIStorageCostComponent(), r.sqlMIBackupCostComponent())
+	}
```

Comment:
For this and Long Term retention we want to add the cost component even if the usage isn't specified. This allows users to discover the usage more easily and see the costs associated with it. e.g:

```
 └─ Long-term retention                                      Monthly cost depends on usage: $0.05 per GB    
```

You'll need to modify the cost component methods to pass in a nil value to the monthly quantity, e.g:

```
	var retention *decimal.Decimal
	if r.LongTermRetentionStorageGB != nil {
		retention = decimalPtr(decimal.NewFromInt(*r.LongTermRetentionStorageGB))
	}

       ...

       MonthlyQuantity: rentention,
```

Reply:
You need to remove the nil check here e.g just:

```go
costComponents = append(costComponents, r.sqlMIStorageCostComponent(), r.sqlMIBackupCostComponent())
```

and then modify `sqlMIBackupCostComponent` so that we pass in the same default value as you've done with the `LongTermRetentionStorageGB`